File: Leetcode_Code/107. Binary Tree Level Order Traversal II/107_bfs.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) return result;
        Queue<TreeNode> Q = new LinkedList<>();
        Q.add(root);
        while (!Q.isEmpty()) {
            int size = Q.size();
            List<Integer> layerElement = new ArrayList<>();
            for (int i = 0; i < size; ++i) {
                TreeNode head = Q.poll();
                layerElement.add(head.val);
                if (head.left != null) Q.add(head.left);
                if (head.right != null) Q.add(head.right);
            }
            result.add(0, layerElement);
        }
        return result;
    }
}

File: Leetcode_Code/107. Binary Tree Level Order Traversal II/107_bfs.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode *root) {
        vector<vector<int>> result;
        if (!root) return result;
        queue<TreeNode *> Q;
        Q.push(root);
        while (!Q.empty()) {
            int size = (int) Q.size();
            vector<int> layerElement;
            for (int i = 0; i < size; ++i) {
                TreeNode *head = Q.front();
                Q.pop();
                layerElement.push_back(head->val);
                if (head->left) Q.push(head->left);
                if (head->right) Q.push(head->right);
            }
            result.insert(result.begin(), layerElement);
        }
        return result;
    }
};
